src/read.c

```diff
@@ -237,17 +237,35 @@ typedef struct avifSampleTable
     uint32_t allSamplesSize; // If this is non-zero, sampleSizes will be empty and all samples will be this size
 } avifSampleTable;
 
+static void avifSampleTableDestroy(avifSampleTable * sampleTable);
+
 static avifSampleTable * avifSampleTableCreate()
 {
     avifSampleTable * sampleTable = (avifSampleTable *)avifAlloc(sizeof(avifSampleTable));
     memset(sampleTable, 0, sizeof(avifSampleTable));
-    avifArrayCreate(&sampleTable->chunks, sizeof(avifSampleTableChunk), 16);
-    avifArrayCreate(&sampleTable->sampleDescriptions, sizeof(avifSampleDescription), 2);
-    avifArrayCreate(&sampleTable->sampleToChunks, sizeof(avifSampleTableSampleToChunk), 16);
-    avifArrayCreate(&sampleTable->sampleSizes, sizeof(avifSampleTableSampleSize), 16);
-    avifArrayCreate(&sampleTable->timeToSamples, sizeof(avifSampleTableTimeToSample), 16);
-    avifArrayCreate(&sampleTable->syncSamples, sizeof(avifSyncSample), 16);
+    if (avifArrayCreate(&sampleTable->chunks, sizeof(avifSampleTableChunk), 16) != AVIF_RESULT_OK) {
```

Comment:
it seems like we can combine these if conditions with an || and then do the "error" path inside the if block ? This way we can avoid the goto.

For example:
```
if (avifArrayCreate(...) != AVIF_RESULT_OK ||
    avifArrayCreate(...) != AVIF_RESULT_OK ||
    avifArrayCreate(...) != AVIF_RESULT_OK) {
    avifSampleTableDestroy(sampleTable);
    return NULL;
}
return sampleTable;
```

This comment also applies for the rest of the code changes where we try to create several arrays and have to fail if one of them fails.

Reply:
Although I prefer a more modern pattern, I believe keeping a goto everywhere for consistency is clearer. I suggest merging the condition with || for shortness but keeping the goto.

Reply:
Thank you both for the suggestion of merging the conditions with `||`. I wrote this as multiple if statements to make it easier to debug this in a debugger. This style is more common in the libavif codebase. Here is an example, which uses a macro (`CHECK`) to make the code less verbose:

```
#define CHECK(A)               \
    do {                       \
        if (!(A))              \
            return AVIF_FALSE; \
    } while (0)

static avifBool avifParseCleanApertureBoxProperty(avifProperty * prop, const uint8_t * raw, size_t rawLen, avifDiagnostics * diag)
{
    BEGIN_STREAM(s, raw, rawLen, diag, "Box[clap]");

    avifCleanApertureBox * clap = &prop->u.clap;
    CHECK(avifROStreamReadU32(&s, &clap->widthN));    // unsigned int(32) cleanApertureWidthN;
    CHECK(avifROStreamReadU32(&s, &clap->widthD));    // unsigned int(32) cleanApertureWidthD;
    CHECK(avifROStreamReadU32(&s, &clap->heightN));   // unsigned int(32) cleanApertureHeightN;
    CHECK(avifROStreamReadU32(&s, &clap->heightD));   // unsigned int(32) cleanApertureHeightD;
    CHECK(avifROStreamReadU32(&s, &clap->horizOffN)); // unsigned int(32) horizOffN;
    CHECK(avifROStreamReadU32(&s, &clap->horizOffD)); // unsigned int(32) horizOffD;
    CHECK(avifROStreamReadU32(&s, &clap->vertOffN));  // unsigned int(32) vertOffN;
    CHECK(avifROStreamReadU32(&s, &clap->vertOffD));  // unsigned int(32) vertOffD;
    return AVIF_TRUE;
}
```

Reply:
Thank you both for the suggestion of merging the conditions with `||`. I wrote this as multiple if statements to make it easier to debug this in a debugger. This style is more common in the libavif codebase. Here is an example, which uses a macro (`CHECK`) to make the code less verbose:

```
#define CHECK(A)               \
    do {                       \
        if (!(A))              \
            return AVIF_FALSE; \
    } while (0)

static avifBool avifParseCleanApertureBoxProperty(avifProperty * prop, const uint8_t * raw, size_t rawLen, avifDiagnostics * diag)
{
    BEGIN_STREAM(s, raw, rawLen, diag, "Box[clap]");

    avifCleanApertureBox * clap = &prop->u.clap;
    CHECK(avifROStreamReadU32(&s, &clap->widthN));    // unsigned int(32) cleanApertureWidthN;
    CHECK(avifROStreamReadU32(&s, &clap->widthD));    // unsigned int(32) cleanApertureWidthD;
    CHECK(avifROStreamReadU32(&s, &clap->heightN));   // unsigned int(32) cleanApertureHeightN;
    CHECK(avifROStreamReadU32(&s, &clap->heightD));   // unsigned int(32) cleanApertureHeightD;
    CHECK(avifROStreamReadU32(&s, &clap->horizOffN)); // unsigned int(32) horizOffN;
    CHECK(avifROStreamReadU32(&s, &clap->horizOffD)); // unsigned int(32) horizOffD;
    CHECK(avifROStreamReadU32(&s, &clap->vertOffN));  // unsigned int(32) vertOffN;
    CHECK(avifROStreamReadU32(&s, &clap->vertOffD));  // unsigned int(32) vertOffD;
    return AVIF_TRUE;
}
```

Reply:
I would greatly appreciate a library-wide AVIF_CHECK macro for error handling. Unfortunately many but not all places use a goto, some with a "cleanup" label, others with an "error" one. This is past the scope of this PR but unifying all these with a single macro (handling a bool or even better an AVIF_RESULT) would be handy and less verbose, at the expense of having a goto label nearly everywhere.

Reply:
I took a look at the "cleanup" label. It is used in functions that must free resources (clean up) before returning, whether success or failure. The functions modified in this pull request free resources only when returning on error. That's why I named the label "error" in this pull request.

Reply:
We only need to use a `goto` label if we need to perform cleanup actions before returning and we want to share the cleanup actions between multiple code paths. If no cleanup is necessary, then it's simpler to just return from the if statements.